### ISSUE TITLE
[E3] Duda capítulo Boundaries

### ISSUE THREAD

**Issue by camibennett** [franciscogazitua/IIC2113-2024#61]
Hola! No me queda claro si la idea es que en el proyecto creemos una clase distinta para cada vez que usamos una lista (por ejemplo) o si podemos hacer una clase genérica de lista y usarla cada vez que tengamos una lista.

Saludos y gracias de antemano :)

**Comment by RodrigoToroIcarte** [comment]
Depende del caso :)

¿Cómo sería tu lista genérica?

**Comment by camibennett** [comment]
Pensaba agregarle los metodos mas comunes, como and, removedme, clear, etc. Creo que si hiciera una clase para cada tipo de lista repetiría mucho código.

**Comment by RodrigoToroIcarte** [comment]
Claro. Es que parte de la idea de encapsular listas es crear una "_interfaz a la medida_". Por ejemplo, si para una lista en particular no quieres dejar que se puedan remover objetos, entonces no le agregas un método para remover cosas. Si tienes una lista genérica pierdes la opción de hacer una interfaz a la medida.

Por otro lado, la idea también es protegernos de cambios que existan a futuro. Por ejemplo, si a futuro nos damos cuenta que era mejor usar un _set_ que una _list_, solo tenemos que cambiar una clase. Si tienes una lista genérica igual podrías hacer ese cambio, pero será un cambio que afecte a todas las listas que tengas en tu programa.

... por eso es que la respuesta es depende. ¿Qué listas tienes que te gustaría agrupar en una lista genérica?

**Comment by camibennett** [comment]
Tengo listas de effects, units, unitDTO, stat, string, skillDTO y Skill en mi código. Quizás me convendría hacer una clase padre de lista genérica y que las listas de cada tipo hereden de esa?

**Comment by RodrigoToroIcarte** [comment]
¿Pero esas listas las retornas o das como argumento a otras clases? Porque si ya están encapsuladas en sus propias clases (i.e., solo se usan dentro de la clase) puede que no tengas que hacer nada.

**Comment by camibennett** [comment]
En general las listas son atributos de clases y muchas veces accedo a ellas desde otras clases.

**Comment by RodrigoToroIcarte** [comment]
¿A qué te refieres con que accedes a ellas? Onda, son públicas?

**Comment by camibennett** [comment]
Si, tengo algunas estructuras de datos que tienen como atributo a una lista.

**Comment by RodrigoToroIcarte** [comment]
Ya veo. En ese caso sí deberías encapsulas las listas en clases. 

Recuerda que una clase debería tener una sola responsabilidad. Y el manejo de algunas de tus listas claves son una responsabilidad.

Por ejemplo, me parece muy razonable tener una clase `Effects` que maneje una lista de efectos. Así le puedes poner métodos particulares a esa lista, como un `ApplyEffects(...)` que recorra los efectos y los aplique.

Una clase `Team` también me parece razonable. Esa clase encapsularía a tu lista de `units`.

... y así.

La lista de `string` que tienes podría ser mejor que la transformes en un arreglo (y uses `params` si la das como argumento a un método).

El tema es que tendría que ver ejemplos concretos para discutir esto mejor. La lista genérica podría funcionar en casos particulares.

En cuanto al código duplicado, no sé qué tanto código duplicado te quede. Como que todas tendrán un método `Add`, pero será una sola línea de código. Además lo que recibe el `Add` dependerá del tipo de la lista (entonces tampoco será código 100% duplicado). Lo mismo ocurrirá con el `Get`. Un `Remove` probablemente no lo tengas y, de tenerlo, puede que requiera otro nombre. Por ejemplo, en `Team` podrías tener un método para eliminar a las unidades muertas (que es de más alto nivel que tener un método `Remove(int unitId)`).

**Comment by camibennett** [comment]
Perfecto, muchas gracias!
